### PR TITLE
Replaced get_behavior_names and get_behavior_spec with behavior_specs property

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Replaced get_behavior_names and get_behavior_spec with behavior_specs property

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by andrewcoh** [comment]
Making it a property is much cleaner.  Just curious as to the reason for creating `BehaviorMapping`. Is it just to have more explicit naming? From what I can tell it has the same functionality as a dictionary.

**Comment by vincentpierre** [comment]
> Making it a property is much cleaner. Just curious as to the reason for creating BehaviorMapping. Is it just to have more explicit naming? From what I can tell it has the same functionality as a dictionary.

Using Mapping allows me to have this property be read only.

File: gym-unity/gym_unity/tests/test_gym.py [hunk 1]
```diff
@@ -19,7 +20,7 @@ def test_gym_wrapper():
     setup_mock_unityenvironment(
         mock_env, mock_spec, mock_decision_step, mock_terminal_step
     )
-
+    print(mock_env.behavior_specs, len(mock_env.behavior_specs))
```
Comment (chriselion) [inline gym-unity/gym_unity/tests/test_gym.py:23]: nit: remove print

File: ml-agents-envs/mlagents_envs/base_env.py [hunk 2]
```diff
@@ -331,17 +345,16 @@ def close(self) -> None:
...
+    def behavior_specs(self) -> BehaviorMapping:
         """
-        Returns the list of the behavior names present in the environment.
+        Returns a Mapping from behavior names to behavior specs.
         Agents grouped under the same behavior name have the same action and
         observation specs, and are expected to behave similarly in the
         environment.
-        This list can grow with time as new policies are instantiated.
-        :return: the list of agent BehaviorName.
+        Note that this mapping can grow and change as new policies are instantiated.
```
Comment (chriselion) [inline ml-agents-envs/mlagents_envs/base_env.py:356]: "and change" - the spec that a behavior name is associated with should always be the same, right?

Maybe "New keys can be added to this mapping as new policies are instantiated"?

File: ml-agents/mlagents/trainers/simple_env_manager.py [hunk 3]
```diff
@@ -53,9 +53,9 @@ def _reset_env(
     @property
     def external_brains(self) -> Dict[BehaviorName, BrainParameters]:
         result = {}
-        for brain_name in self.env.get_behavior_names():
+        for brain_name in self.env.behavior_specs:
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/simple_env_manager.py:56]:
```suggestion
        for behavior_name, behavior_spec in self.env.behavior_specs.items():
```
(and update line below)

File: ml-agents/mlagents/trainers/subprocess_env_manager.py [hunk 4]
```diff
@@ -131,15 +131,15 @@ def _send_response(cmd_name: EnvironmentCommand, payload: Any) -> None:
...
         all_step_result: AllStepResult = {}
-        for brain_name in env.get_behavior_names():
+        for brain_name in env.behavior_specs:
             all_step_result[brain_name] = env.get_steps(brain_name)
         return all_step_result
 
     def external_brains():
         result = {}
-        for brain_name in env.get_behavior_names():
+        for brain_name in env.behavior_specs:
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/subprocess_env_manager.py:140]: iterate over `items()`

File: ml-agents-envs/mlagents_envs/base_env.py [hunk 5]
```diff
@@ -331,17 +345,16 @@ def close(self) -> None:
         """
         pass
 
+    @property
     @abstractmethod
-    def get_behavior_names(self) -> List[BehaviorName]:
+    def behavior_specs(self) -> BehaviorMapping:
```
Comment (chriselion) [inline ml-agents-envs/mlagents_envs/base_env.py:350]:
```suggestion
    def behavior_specs(self) -> Mapping:
```
I think the abstract class should just return an abstract Mapping. Other implementations could return a plain dict if they wanted.

**Comment by chriselion** [comment]
> Using Mapping allows me to have this property be read only.

Mapping is read-only (as opposed to MutableMapping), so using Mapping as the return type should be sufficient and you could just return `self._env_specs` in UnityEnvironment. It wouldn't be enforced at runtime, just by type checking and IDE syntax highlighting (but a user determined to mess things up could modify `BehaviorMapping _dict` too).

**Comment by vincentpierre** [comment]
> Mapping is read-only (as opposed to MutableMapping), so using Mapping as the return type should be sufficient and you could just return self._env_specs in UnityEnvironment. It wouldn't be enforced at runtime, just by type checking and IDE syntax highlighting (but a user determined to mess things up could modify BehaviorMapping _dict too).

Would you recommend dropping the `BehaviorMapping` alltogether?
I think it is safer to do it this way, it makes it harder to mess up (while not impossible) which is better in my opinion.

**Comment by chriselion** [comment]
I think BehaviorMapping is overkill, but I'm not strongly opposed to it. I think having `Mapping` as part of the interface is the important part.

File: ml-agents-envs/mlagents_envs/base_env.py [hunk 6]
```diff
@@ -331,17 +345,16 @@ def close(self) -> None:
...
+    def behavior_specs(self) -> Mapping:
         """
-        Returns the list of the behavior names present in the environment.
+        Returns a Mapping from behavior names to behavior specs.
         Agents grouped under the same behavior name have the same action and
         observation specs, and are expected to behave similarly in the
         environment.
-        This list can grow with time as new policies are instantiated.
-        :return: the list of agent BehaviorName.
+        Note that new keys can be added to this mapping as new policies areinstantiated.
```
Comment (chriselion) [inline ml-agents-envs/mlagents_envs/base_env.py:356]:
```suggestion
        Note that new keys can be added to this mapping as new policies are instantiated.
```

File: gym-unity/gym_unity/envs/__init__.py [hunk 7]
```diff
@@ -49,7 +49,7 @@ def __init__(
         self._env = unity_env
 
         # Take a single step so that the brain information will be sent over
-        if not self._env.get_behavior_names():
+        if len(self._env.behavior_specs) == 0:
```
Comment (chriselion) [inline gym-unity/gym_unity/envs/__init__.py:52]: nit: pretty sure you can do `if not self._env.behavior_specs`

**Comment by vincentpierre** [comment]
> Looks good. Any documentation/examples that need to change too?

Yes, coming soon

**Comment by vincentpierre** [comment]
> Looks good. Any documentation/examples that need to change too?

@chriselion Made the documentation changes.

**Pull request merged**